app/models/task.rb

```diff
@@ -5,9 +5,17 @@ class Task < ApplicationRecord
     state :unstarted, initial: true
     state :in_progress
     state :completed
+    state :correcting
 
     event :completed do
-      transitions from: %i[unstarted in_progress completed], to: :completed
+      transitions from: %i[unstarted in_progress correcting completed], to: :completed
+    end
+
+    event :cancel_correction do
+      before do
+        destroy_incomplete_correction_submissions if correcting?
```

Comment:
So this `if` made me think this shouldn't be done as a `before`. Looking at the [aasm docs](https://github.com/aasm/aasm/blob/master/README.md#callbacks), we should probably use an `after` callback instead:

```ruby
transitions from: :correcting, to: :completed, after: destroy_incomplete_correction_submissions
```

`

Reply:
The documentation is not particularly informative on this point: if we do the destructive actions in the `after` callback, and the destructive action errors thus rolling the transaction back, it looks like the state transition won't be undone. So we'd end up with a `completed` task with the associated correction files not deleted.

Reply:
> So we'd end up with a completed task with the associated correction files not deleted.

That would be far from ideal! More reason for some solid test coverage on this event to make sure we understand and verify how we want it to behave under happy and sad paths.

Reply:
So I've added the spec as above. However, I still can't find a way of doing this with aasm without that `if`, and I'm not *too* unhappy about it. Ideally, I'd like a callback that happens after the check to see if this is a valid transition, but I haven't been able to find it. If I try something like `before_exit` which that documentation *seems* to suggest, everything just blows up.